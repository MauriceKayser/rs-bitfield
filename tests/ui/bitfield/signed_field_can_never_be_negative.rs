//! Assertion failure: _SIGNED_TYPE_IN_FIELD_0_CAN_NEVER_BE_NEGATIVE

extern crate alloc;

#[bitfield::bitfield(16)]
struct BitField(#[field(1, 7)] Field); // Needs a size of 8 instead of 7 to store negative values.

#[derive(Debug, bitfield::Field)]
#[repr(i8)] // Alternatively, use `u8` instead.
enum Field {
    FMinus1 = -1,
    F1 = 1,
    F2,
    F3
}

fn main() {}