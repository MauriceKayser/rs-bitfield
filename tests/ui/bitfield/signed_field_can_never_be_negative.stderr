error[E0080]: evaluation of constant value failed
 --> tests/ui/bitfield/signed_field_can_never_be_negative.rs:6:28
  |
6 | struct BitField(#[field(1, 7)] Field); // Needs a size of 8 instead of 7 to store negative values.
  |                            ^ the evaluated program panicked at 'Signed type can never be negative', $DIR/tests/ui/bitfield/signed_field_can_never_be_negative.rs:6:28
  |
  = note: this error originates in the macro `$crate::panic::panic_2021` which comes from the expansion of the macro `panic` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0271]: type mismatch resolving `<Field as TryFrom<u8>>::Error == u8`
 --> tests/ui/bitfield/signed_field_can_never_be_negative.rs:6:32
  |
6 | struct BitField(#[field(1, 7)] Field); // Needs a size of 8 instead of 7 to store negative values.
  |                                ^^^^^ expected `u8`, found `Infallible`

error[E0277]: the trait bound `Field: TryFrom<u8>` is not satisfied
 --> tests/ui/bitfield/signed_field_can_never_be_negative.rs:6:28
  |
6 | struct BitField(#[field(1, 7)] Field); // Needs a size of 8 instead of 7 to store negative values.
  |                            ^^^^-----
  |                            |   |
  |                            |   required by a bound introduced by this call
  |                            the trait `From<u8>` is not implemented for `Field`
  |
  = help: the trait `TryFrom<u8>` is not implemented for `Field`
          but trait `TryFrom<i8>` is implemented for it
  = help: for that trait implementation, expected `i8`, found `u8`
  = note: required for `u8` to implement `Into<Field>`
  = note: required for `Field` to implement `TryFrom<u8>`

error[E0277]: the trait bound `Field: From<u8>` is not satisfied
 --> tests/ui/bitfield/signed_field_can_never_be_negative.rs:6:32
  |
6 | struct BitField(#[field(1, 7)] Field); // Needs a size of 8 instead of 7 to store negative values.
  |                                ^^^^^ the trait `From<u8>` is not implemented for `Field`
  |
  = help: the trait `TryFrom<u8>` is not implemented for `Field`
          but trait `TryFrom<i8>` is implemented for it
  = help: for that trait implementation, expected `i8`, found `u8`
  = note: required for `u8` to implement `Into<Field>`
  = note: required for `Field` to implement `TryFrom<u8>`
